### ISSUE TITLE
如何用自己的数据做预训练？

### ISSUE THREAD

**Issue by sz547073165** [PaddlePaddle/ERNIE#353]
看到这个[预训练 (ERNIE 1.0)](https://github.com/PaddlePaddle/ERNIE/blob/develop/README.zh.md#预训练-ernie-10)想用自己的数据进行预训练

执行了这个脚本[script/zh_task/pretrain.sh](https://github.com/PaddlePaddle/ERNIE/blob/develop/script/zh_task/pretrain.sh)

这个脚本跑不起来 报错
`[root@localhost zh_task]# sh ./pretrain.sh 
+ export FLAGS_eager_delete_tensor_gb=0
+ FLAGS_eager_delete_tensor_gb=0
+ export FLAGS_sync_nccl_allreduce=1
+ FLAGS_sync_nccl_allreduce=1
+ export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
+ CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
++ hostname -i
+ python3 /root/ERNIE/pretrain_launch.py --nproc_per_node 8 --selected_gpus 0,1,2,3,4,5,6,7 --node_ips ::1 127.0.0.1 --node_id 0 ./train.py --use_cuda True --is_distributed False --use_fast_executor True --weight_sharing True --in_tokens true --batch_size 8192 --vocab_path ./config/vocab.txt --train_filelist ./data/train_filelist --valid_filelist ./data/valid_filelist --validation_steps 100 --num_train_steps 1000000 --checkpoints ./checkpoints --save_steps 10000 --ernie_config_path ./config/ernie_config.json --learning_rate 1e-4 --use_fp16 false --weight_decay 0.01 --max_seq_len 512 --skip_steps 10
usage: pretrain_launch.py [-h] [--ernie_config_path ERNIE_CONFIG_PATH]
                          [--init_checkpoint INIT_CHECKPOINT]
                          [--checkpoints CHECKPOINTS]
                          [--weight_sharing WEIGHT_SHARING]
                          [--generate_neg_sample GENERATE_NEG_SAMPLE]
                          [--epoch EPOCH] [--learning_rate LEARNING_RATE]
                          [--lr_scheduler {linear_warmup_decay,noam_decay}]
                          [--weight_decay WEIGHT_DECAY]
                          [--num_train_steps NUM_TRAIN_STEPS]
                          [--warmup_steps WARMUP_STEPS]
                          [--save_steps SAVE_STEPS]
                          [--validation_steps VALIDATION_STEPS]
                          [--use_fp16 USE_FP16]
                          [--use_dynamic_loss_scaling USE_DYNAMIC_LOSS_SCALING]
                          [--init_loss_scaling INIT_LOSS_SCALING]
                          [--incr_every_n_steps INCR_EVERY_N_STEPS]
                          [--decr_every_n_nan_or_inf DECR_EVERY_N_NAN_OR_INF]
                          [--incr_ratio INCR_RATIO] [--decr_ratio DECR_RATIO]
                          [--skip_steps SKIP_STEPS] [--verbose VERBOSE]
                          [--train_filelist TRAIN_FILELIST]
                          [--valid_filelist VALID_FILELIST]
                          [--test_filelist TEST_FILELIST]
                          [--vocab_path VOCAB_PATH]
                          [--max_seq_len MAX_SEQ_LEN]
                          [--batch_size BATCH_SIZE] [--in_tokens IN_TOKENS]
                          [--is_distributed IS_DISTRIBUTED]
                          [--use_cuda USE_CUDA]
                          [--use_fast_executor USE_FAST_EXECUTOR]
                          [--do_test DO_TEST]
pretrain_launch.py: error: unrecognized arguments: --node_id 0 ./train.py
`

**Comment by Meiyim** [comment]
看起来你这边`$(hostname -i)`返回的结果格式不太正确。你可以手动指定您的本机ip，如下
```script
...
--node_ips 127.0.0.1 \
...
```
或者你可以尝试本地运行（而非分布式运行）：
只需去除`pretrain_launch.py`这一段，把启动脚本改为：
```script
python -u ./ernie/train.py --use_cuda True \
                --is_distributed False\
                --use_fast_executor True \
                --weight_sharing True \
                --in_tokens true \
                --batch_size 8192 \
                --vocab_path ./config/vocab.txt \
                --train_filelist ./data/train_filelist \
                --valid_filelist ./data/valid_filelist \
                --validation_steps 100 \
                --num_train_steps 1000000 \
                --checkpoints ./checkpoints \
                --save_steps 10000 \
                --ernie_config_path ./config/ernie_config.json \
                --learning_rate 1e-4 \
                --use_fp16 false \
                --weight_decay 0.01 \
                --max_seq_len 512 \
                --skip_steps 10
```

**Comment by woyijkl1** [comment]
“seg_labels”是利用百度内部工具得到的？  
那请问非百度开发者如何获得seglabel

**Comment by Meiyim** [comment]
seg_labels仅仅是一个包含切词边界信息的array。
举个例子，您可以利用jieba分词：
例子 `jieba.cut("今天吃了么"） => ["今天", "吃" ,"了", "么"]`

则输入text明文：["[CLS]", "今", 天", "吃" ,"了", "么", "[SEP]"]
text_id:       [1, 2, 3, 4, 5, 6,  7]
seg_labels: [-1, 0, 1, 0, 0, 0,-1]

**Comment by woyijkl1** [comment]
ok 十分感谢

**Comment by sz547073165** [comment]
官方有提供数据预处理脚本不

**ernie.reader.task_reader.BaseReader._convert_example_to_record**

这个方法处理的数据，好像没seg_labels，多了个“qid”

`
Record = namedtuple('Record', [
                'token_ids', 'text_type_ids', 'position_ids', 'label_id', 'qid'
            ])
`

如：

`
text_a = '预包装食品、化妆品、日用品及卫生用品经销。'
label = 'F'
`

`
Record(token_ids=[1, 695, 392, 371, 494, 100, 6, 73, 2078, 100, 6, 139, 29, 100, 107, 573, 21, 29, 100, 60, 579, 12043, 2], text_type_ids=[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], position_ids=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22], label_id='5', qid=None)
`

**Comment by sz547073165** [comment]
数据预处理的脚本，这里找到有一个，看样子是根据google bert提供的数据处理脚本改写的：

> https://github.com/wipen/bert_and_ernie/blob/master/create_ernie_pretraining_data.py

不过line233处，生成seg_labels数组的逻辑有问题，可以根据[这个逻辑](https://github.com/PaddlePaddle/ERNIE/issues/353#issuecomment-546920929)自己改一下，还有别的地方要改的忘记了，需要的童鞋自己debug一下。

**Comment by hrwise-nlp** [comment]
请问下我这边运行一直卡在log那里是什么原因？
我check了一下log，发现如下问题
server not ready, wait 3 sec to retry...
